refactor(echo): introduce declaring workflow

denis-kralj-novu opened the pull request:
### What changed? Why was the change needed?
- Add workflow function
- Adjust tests to work with new paradigm
- Extract independent tests to separate suite
- Adjust handler and serve methods

### Enterprise PR dependence

[✅ MERGED] This PR depends on [this EE PR](https://github.com/novuhq/packages-enterprise/pull/145)

denis-kralj-novu committed:
refactor(echo): introduce declaring workflow

Add workflow function
Adjust tests to work with new paradigm
Extract independent tests to separate suite
Adjust handler and serve methods

denis-kralj-novu committed:
refactor(echo): introduce declaring workflow

Add missed serve adjustments

rifont commented on packages/echo/src/express.ts:
```diff
@@ -8,9 +9,11 @@ import { type SupportedFrameworkName } from './types';
 export const frameworkName: SupportedFrameworkName = 'express';
 
 export const serve = (options: ServeHandlerOptions): any => {
+  const client = options.client ? options.client : new Echo();
```
I think we can handle this instantiation inside the `EchoRequestHandler`, the `...options` spread will ensure the client is injected if available, and we won't need to modify any framework files.

denis-kralj-novu replied:
Yep, much better approach

denis-kralj-novu committed:
refactor(echo): introduce declaring workflow

PR comments

denis-kralj-novu committed:
Merge branch 'next' into nv-3888-refactor-code-first-sdk-interface-with-support-for-declaring

denis-kralj-novu commented on packages/echo/src/handler.ts:
```diff
@@ -64,9 +66,10 @@ export class EchoRequestHandler<Input extends any[] = any[], Output = any> {
     this.handler = options.handler;
-    this.client = options.client;
+    this.client = options.client ? options.client : new Echo();
+    this.client.addWorkflows(options.workflow);
```
@rifont do we concern ourselves that devs might use this method as well? We don't say it it in our docs (note to self: update readme with how to use the framework) but there is still a chance devs might see this and use it

SokratisVidros replied:
Do we have to expose it? I see it's public, but we can make it private.

denis-kralj-novu replied:
if I make it private then it is not accessible in the handler either. I can move all the logic associated with persisting the workflows in the handler but it is more work.

Is there a way in TS to make a class method available within the package but not outside of it?

SokratisVidros replied:
I see. Maybe `this.client.workflows = workflows` then and use getter setters?

denis-kralj-novu replied:
let's not worry about that now. The docs talk about how we want to use it so if a developer uses the other methods without understanding how it works it is on them. We have provided a good DX, but we cannot force peoples hand. We can circle back to this if it becomes a problem

rifont replied:
With the switch to implicitly instantiate the client on `serve`, I think most Developers _won't_ see the client methods anymore, I'd expect less than 5% will need to explicitly use the `Client`.

We could take a good variety of approaches to not expose it, but I also like the idea of the Framework providing escape hatches for Developers to build custom things on top of. So long as we choose appropriate naming, which I believe `addWorkflows` is, we should grant Developers flexibility to go wild with this. We might learn something from their creativity that we can build out as a first-class capability later.

SokratisVidros commented on packages/echo/src/workflow.ts:
```diff
@@ -0,0 +1,203 @@
+        return acc;
+      }, {} as Record<ChannelStepEnum, ActionStep<any, any>>),
+      /*
+       * Temporary workaround for inApp, which has snake_case step type
```
Is the `in_app` customer facing? If it is, let's change it to `inApp` and internally change it again until we fix the API.

denis-kralj-novu replied:
added to my todo, won't target it in this PR

rifont replied:
`inApp` is the customer facing interface here, no further action needed.

SokratisVidros commented on packages/echo/src/client.ts:
```diff
@@ -72,15 +72,26 @@ export class Echo {
+  public addWorkflows(workflows: Array<DiscoverWorkflowOutput>) {
+    for (const workflow of workflows) {
+      if (this.discoveredWorkflows.some((existing) => existing.workflowId === workflow.workflowId)) {
+        throw new WorkflowAlreadyExistsError(workflow.workflowId);
```
Should we throw an error or discard duplicates? What's better DX in this case?

denis-kralj-novu replied:
I don't think implicitly discarding duplicates it the way to go. How would we know which copy is the "correct" one? the one defined first or last? Throwing explains to the user that the identifiers have to be unique

SokratisVidros replied:
Thanks for the clarification.

SokratisVidros commented on packages/echo/src/client.ts:
```diff
@@ -72,15 +72,26 @@ export class Echo {
+    this.apiKey = config?.apiKey ?? process.env.NOVU_API_KEY;
     this.backendUrl = config?.backendUrl ?? DEFAULT_NOVU_API_BASE_URL;
-    this.devModeBypassAuthentication = config?.devModeBypassAuthentication || false;
+    this.devModeBypassAuthentication = config?.devModeBypassAuthentication || process.env.NODE_ENV === 'development';
```
```suggestion
    this.strictAuthentication = config?.strictAuthentication || process.env.NODE_ENV === 'development';
```

denis-kralj-novu replied:
if I understand correctly, your name adjustment means we are checking the inverse of what was there before (should be bypass auth because dev vs. enable strict auth). that would mean that the `process.env.NODE_ENV === 'development'` should also be the opposite

denis-kralj-novu replied:
updated

SokratisVidros commented on packages/echo/src/client.ts:
```diff
@@ -72,15 +72,26 @@ export class Echo {
   constructor(config?: ClientConfig) {
-    this.apiKey = config?.apiKey;
+    // TODO: transform this as per notion DX guide
+    this.apiKey = config?.apiKey ?? process.env.NOVU_API_KEY;
```
I suggest moving the config(options) parsing to a separate method that will handle the default, env var fallback logic in one place and enforce naming convention on options and env vars.

denis-kralj-novu replied:
done

SokratisVidros commented on packages/echo/src/client.ts:
```diff
@@ -72,15 +72,26 @@ export class Echo {
-    this.apiKey = config?.apiKey;
+    // TODO: transform this as per notion DX guide
+    this.apiKey = config?.apiKey ?? process.env.NOVU_API_KEY;
     this.backendUrl = config?.backendUrl ?? DEFAULT_NOVU_API_BASE_URL;
```
```suggestion
    this.apiUrl = config?.apiUrl ?? NOVU_API_URL;
```

denis-kralj-novu replied:
updated

SokratisVidros commented on packages/echo/src/client.ts:
```diff
@@ -72,15 +72,26 @@ export class Echo {
   public static NOVU_SIGNATURE_HEADER = HttpHeaderKeysEnum.SIGNATURE;
 
   constructor(config?: ClientConfig) {
```
```suggestion
  constructor(options?: ClientOptions) {
```

denis-kralj-novu replied:
✔️

denis-kralj-novu committed:
refactor(echo): introduce declaring workflow

Add tests for string util method
Build configuration based on DX guide

SokratisVidros commented on packages/echo/src/client.ts:
```diff
@@ -72,15 +72,47 @@ export class Echo {
+  public addWorkflows(workflows: Array<DiscoverWorkflowOutput>) {
+    for (const workflow of workflows) {
+      if (this.discoveredWorkflows.some((existing) => existing.workflowId === workflow.workflowId)) {
+        throw new WorkflowAlreadyExistsError(workflow.workflowId);
```
Having said that, instead of an error class per error type it's better DX to have a NovuFrameworkError and add messages and reason. That way it's easier for consumers to consume a single type and for us it creates less code.

denis-kralj-novu replied:
you think that if the user wants to react on specific error states that matching message content is the way to go? in my experience, we do a catch all to log and then re throw, while particular errors that interest us are better targeted by matching types as opposed to message content

SokratisVidros replied:
In TS, every new type is an extra import. So the catch-all code becomes cumbersome. My suggestion wasn't just a message. It combines message, reason, and code in a `NovuError` class. The code is what helps you dictate logic. Let me know if this helps.

denis-kralj-novu replied:
yep, that makes more sense.

denis-kralj-novu replied:
Ok so I took the time to look at the situation more. Basically all errors inherit up the tree the `NovuError` base class, and this class includes a http status code as well as a code that describes what happened from the framework perspective.

GIven how try/catch works, with this inheritance approach the user can aim for exactly the specific error type they want with `instanceof`. If they want to work with multiple error type imports and handle each case that way, they can. On the other hand, they can simply aim for `NovuError` and see its code to decide what to do next.

Does this make sense @SokratisVidros ?

rifont replied:
Exactly right @denis-kralj-novu , the `NovuError` is a generic Error interface, the specific Errors are just sugar for explicitly declaring errors throughout the framework with the `code` embedded. Some errors also have some sugar in the constructor to provide a nice error message where they are used in multiple places, to ensure consistent formatting.

In a future iteration, we can add an [Addressable Error](https://rauchg.com/2016/addressable-errors) to each of the error classes for better DX.

This approach grants SDK consumers the flexibility to handle different types of errors with `instanceof`, whilst also granting flexibility to use the `NovuError.code` to discriminate on the error if their preference is for error object access instead.

denis-kralj-novu committed:
refactor(echo): introduce declaring workflow

Remove dangling constant

denis-kralj-novu committed:
refactor(echo): introduce declaring workflow

Fix testing

denis-kralj-novu committed:
Merge branch 'next' into nv-3888-refactor-code-first-sdk-interface-with-support-for-declaring

denis-kralj-novu committed:
refactor(echo): introduce declaring workflow

Update .source

denis-kralj-novu committed:
refactor(echo): introduce declaring workflow

Update .source

denis-kralj-novu committed:
refactor(echo): introduce declaring workflow

Rename class Echo to Client

denis-kralj-novu committed:
refactor(echo): introduce declaring workflow

PR comments

denis-kralj-novu committed:
refactor(echo): introduce declaring workflow

Fix auth bug

denis-kralj-novu committed:
Merge branch 'next' into nv-3888-refactor-code-first-sdk-interface-with-support-for-declaring

SokratisVidros commented on packages/echo/src/client.test.ts:
```diff
@@ -1,98 +1,39 @@
       }));
     });
+
+    echo = new Client();
```
```suggestion
    client = new Client();
```

denis-kralj-novu replied:
fixed here and file wide

SokratisVidros commented on packages/echo/src/client.ts:
```diff
@@ -71,16 +71,48 @@ export class Echo {
-    this.backendUrl = config?.backendUrl ?? DEFAULT_NOVU_API_BASE_URL;
-    this.devModeBypassAuthentication = config?.devModeBypassAuthentication || false;
+  constructor(options?: ClientOptions) {
+    const builtConfiguration = this.buildConfiguration(options);
```
🤓 

```suggestion
    const opts = this.buildOptions(options);
```

denis-kralj-novu replied:
good eye

denis-kralj-novu committed:
Merge branch 'next' into nv-3888-refactor-code-first-sdk-interface-with-support-for-declaring

denis-kralj-novu committed:
refactor(echo): introduce declaring workflow

Refactor tests

denis-kralj-novu committed:
Merge branch 'next' into nv-3888-refactor-code-first-sdk-interface-with-support-for-declaring

denis-kralj-novu committed:
refactor(echo): introduce declaring workflow

PR comments
Rename some instances of echo

denis-kralj-novu committed:
refactor(echo): introduce declaring workflow

Fix test

denis-kralj-novu committed:
refactor(echo): introduce declaring workflow

Get pipeline to only run ee tests when ee flag is on

denis-kralj-novu commented on .github/workflows/reusable-api-e2e.yml:
```diff
@@ -90,6 +90,7 @@ jobs:
         run: CI='' pnpm build:api --skip-nx-cache
 
       - name: Run E2E tests
+        if: ${{ !inputs.ee }}
```
don't run these tests if we are only interested in the ee tests

rifont replied:
Nice! Been meaning to address this one 🙏

denis-kralj-novu committed:
refactor(echo): introduce declaring workflow

More renaming

denis-kralj-novu committed:
Merge branch 'next' into nv-3888-refactor-code-first-sdk-interface-with-support-for-declaring

denis-kralj-novu commented on packages/echo/src/handler.ts:
```diff
@@ -23,16 +23,18 @@ import {
   frameworkName: string;
-  client: Echo;
+  client?: Client;
+  workflow: Array<DiscoverWorkflowOutput>;
```
I've seen this in the discussion in slack and in the DX guide. any particular reason why we are using a singular for a property that is containing an array?

rifont replied:
Good catch, that was a typo! It should be plural `workflows`

rifont replied:
```suggestion
  workflows: Array<DiscoverWorkflowOutput>;
```

denis-kralj-novu committed:
refactor(echo): introduce declaring workflow

Fix broken test

denis-kralj-novu committed:
refactor(echo): introduce declaring workflow

Rename option props per PR request

denis-kralj-novu committed:
Merge branch 'next' into nv-3888-refactor-code-first-sdk-interface-with-support-for-declaring

denis-kralj-novu committed:
refactor(echo): introduce declaring workflow

Rename missed properties
Fix bug with bool flag parsing

denis-kralj-novu committed:
refactor(echo): introduce declaring workflow

Rename BaseError to NovuError

rifont commented on apps/api/e2e/echo.server.ts:
```diff
@@ -2,22 +2,22 @@ import * as http from 'http';
   }
 
-  async start() {
+  async start(workflows: Array<DiscoverWorkflowOutput> = []) {
```
<!-- Suggestions propose improvements to the current subject. It's important to be explicit and clear on what is being suggested and why it is an improvement. Consider using patches and the blocking or non-blocking decorations to further communicate your intent.  -->
✏ **suggestion:** I'd suggest making this an object containing `{workflows}` as we'll soon add more domains.‏

denis-kralj-novu replied:
done

rifont commented on packages/create-novu-app/templates/app-react-email/ts/app/echo/client.ts:
```diff
@@ -5,7 +5,7 @@ export const echo = new Echo({
    * Enable this flag only during local development
    */
-  devModeBypassAuthentication: process.env.NODE_ENV === "development",
+  strictAuthentication: process.env.NODE_ENV !== "development",
```
<!-- Praises highlight something positive. Try to leave at least one of these comments per review. Do not leave false praise (which can actually be damaging). Do look for something to sincerely praise.  -->
👏 **praise:** Nice, much more readable.‏

rifont commented on packages/echo/src/client.ts:
```diff
@@ -57,33 +57,75 @@ JSONSchemaFaker.option({
+    }, '');
+
+    return builtConfiguration;
+  }
```
The client is getting a bit bloated with "useless" methods, methods that don't actually need to access instance properties (i.e. `this.<property>`) and can therefore be refactored to stateless functions.

I don't think we should action now, but let's keep this in mind for future refactoring - we should be able to reduce the size of the `Client` quite significantly.

rifont commented on packages/echo/src/client.ts:
```diff
@@ -57,33 +57,75 @@ JSONSchemaFaker.option({
+      strictAuthentication: true,
+    };
+
+    // boolean flag risks getting parsed wrong in Object.entries, so this is to compensate
+    if (providedOptions && typeof providedOptions.strictAuthentication === 'boolean') {
+      builtConfiguration.strictAuthentication = providedOptions.strictAuthentication;
+    } else if (process.env.NOVU_STRICT_AUTHENTICATION === 'false') {
+      builtConfiguration.strictAuthentication = false;
+    }
+
+    Object.entries(builtConfiguration).reduce((_acc, [prop]) => {
+      if (prop === 'strictAuthentication') {
+        return '';
+      }
+      const envVar = `NOVU_${toConstantCase(prop)}`;
+      if (providedOptions && providedOptions[prop]) {
+        builtConfiguration[prop] = providedOptions[prop];
+      } else if (process.env[envVar]) {
+        builtConfiguration[prop] = process.env[envVar];
+      }
+
+      return '';
+    }, '');
```
<!-- Suggestions propose improvements to the current subject. It's important to be explicit and clear on what is being suggested and why it is an improvement. Consider using patches and the blocking or non-blocking decorations to further communicate your intent.  -->
✏ **suggestion:** Readability is suffering a bit here. Give the small number of config options today, I'll suggest we simplify this to a more classical style of `if (config.option !== undefined) { // handle option parsing/setting }`. We'll seek to use Zod across this codebase eventually and we can replace with a Zod schema with built-in parsing later. Wdyt?‏

denis-kralj-novu replied:
agree that it is an eyesore, The suggestion is valid and I was struggling with myself if to just do it that way :)

rifont commented on packages/echo/src/utils/string.utils.test.ts:
```diff
@@ -0,0 +1,20 @@
+import { toConstantCase } from './string.utils';
+
+describe('convert to snake case', () => {
```
```suggestion
describe('convert to constant case', () => {
```

rifont reviewed (approved):
Awesome work! Left a few comments, most don't need to be addressed right now but are for future reference.

denis-kralj-novu committed:
refactor(echo): introduce declaring workflow

Pluralize workflow prop

denis-kralj-novu committed:
refactor(echo): introduce declaring workflow

PR suggestion

denis-kralj-novu committed:
refactor(echo): introduce declaring workflow

Clear legacy workflow function

denis-kralj-novu committed:
refactor(echo): introduce declaring workflow

Refactor options parsing

denis-kralj-novu committed:
refactor(echo): introduce declaring workflow

PR suggestion

denis-kralj-novu committed:
Merge branch 'next' into nv-3888-refactor-code-first-sdk-interface-with-support-for-declaring

denis-kralj-novu committed:
refactor(echo): introduce declaring workflow

Make workflow method sync
Remove awaiting
Update sync and diff endpoints to new controller
Update header key values

denis-kralj-novu committed:
Merge branch 'next' into nv-3888-refactor-code-first-sdk-interface-with-support-for-declaring

denis-kralj-novu committed:
refactor(echo): introduce declaring workflow

Update .source reference

denis-kralj-novu committed:
Merge branch 'next' into nv-3888-refactor-code-first-sdk-interface-with-support-for-declaring

denis-kralj-novu committed:
refactor(echo): introduce declaring workflow

Update .source reference

denis-kralj-novu committed:
refactor(echo): introduce declaring workflow

Update .source reference

denis-kralj-novu committed:
refactor(echo): introduce declaring workflow

Remove awaiting of the workflow function

denis-kralj-novu commented on packages/echo/src/client.test.ts:
```diff
@@ -166,60 +107,65 @@ describe('Echo Client', () => {
+      client.addWorkflows([newWorkflow]);
+
+      // wait for discovery to finish
+      await new Promise((resolve) => setTimeout(resolve, 1));
```
@rifont so here is a problem with the non awaited workflow method that needs to be resolved in a timely manner. Because this test defines a pretty complex workflow, not all the async methods have time to finish before the `discover()` method below is called. Waiting even 1 millisecond 'fixes' this. 

I've confirmed that this is the case in next as well with the old way of defining workflows. If I remove the await from the `client.workflow()` calls, I still need this wait time to have this particular test pass.

I've tried to play around with the code but no real resolution came to my mind. In reality, the user can do operations in the workflow definition that take an arbitrary amount of time, and discovery won't pick up depending steps until those promises are fulfilled.

How would we target this?

denis-kralj-novu committed:
refactor(echo): introduce declaring workflow

Update .source reference

scopsy committed:
Merge branch 'next' into nv-3888-refactor-code-first-sdk-interface-with-support-for-declaring

denis-kralj-novu committed:
refactor(echo): introduce declaring workflow

Update .source reference

denis-kralj-novu committed:
Merge branch 'next' into nv-3888-refactor-code-first-sdk-interface-with-support-for-declaring
